pyteal/ast/multi.py

```diff
@@ -0,0 +1,79 @@
+from typing import Callable, List, Union, TYPE_CHECKING
+
+from ..types import TealType
+from ..ir import TealOp, Op, TealBlock
+from .expr import Expr
+from .leafexpr import LeafExpr
+from .scratch import ScratchSlot
+from .seq import Seq
+
+if TYPE_CHECKING:
+    from ..compiler import CompileOptions
+
+
+class MultiValue(LeafExpr):
+    """Represents an operation that returns more than one value"""
+
+    def __init__(
+        self,
+        op: Op,
+        types: List[TealType],
+        *,
+        immediate_args: List[Union[int, str]] = None,
+        args: List[Expr] = None
+    ):
+        """Create a new MultiValue.
+
+        Args:
+            op: The operation that returns values.
+            types: The types of the returned values.
+            immediate_args (optional): Immediate arguments for the op. Defaults to None.
+            args (optional): Stack arguments for the op. Defaults to None.
+        """
+        super().__init__()
+        self.op = op
+        self.types = types
+        self.immediate_args = immediate_args if immediate_args is not None else []
+        self.args = args if args is not None else []
+
+        self.output_slots = [ScratchSlot() for _ in self.types]
+
+    def outputReducer(self, reducer: Callable[..., Expr]) -> Expr:
+        input = [slot.load(self.types[i]) for i, slot in enumerate(self.output_slots)]
+        return Seq(self, reducer(*input))
+
+    def __str__(self):
+        ret_str = "(({}".format(self.op)
+        for a in self.immediate_args:
+            ret_str += " " + a.__str__()
+
+        for a in self.args:
+            ret_str += " " + a.__str__()
+        ret_str += ") "
+
+        ret_str += " ".join([slot.store().__str__() for slot in self.output_slots])
+        ret_str += ")"
+
+        return ret_str
+
+    def __teal__(self, options: "CompileOptions"):
+        tealOp = TealOp(self, self.op, *self.immediate_args)
+        callStart, callEnd = TealBlock.FromOp(options, tealOp, *self.args)
+
+        curEnd = callEnd
+        # the list is reversed in order to preserve the ordering of the opcode's returned
+        # values. ie the output to stack [A, B, C] should correspond to C->output_slots[2]
+        # B->output_slots[1], and A->output_slots[0].
+        for slot in reversed(self.output_slots):
```

Comment:
I would like to ask if there's possibility to improve the performance of these lines.

The process of "reading an global state `EX_APP_KEY` of app `ID` then assert" in TEAL was like this:

```teal
int ID
byte "EX_APP_KEY"
app_global_get_ex
assert
```

But in PyTeal, with this MultiValue, the generated code becomes:

```teal
int ID
byte "EX_APP_KEY"
app_global_get_ex
store 1
store 0
load 1
assert
load 0
```

The OpCost of this process was doubled from 4 to 8 (Or 2 to 6, if the two arguments are already on the stack). In other words, the performance was reduced by half or a third. 



I want to [PR](https://github.com/algorand/pyteal/pull/541) for `addw` and `mulw`. But using this seems would change the OpCost from 1 (in TEAL) of `addw` to 5 (in PyTeal) of these following lines:

```teal
addw
store 0
store 1
load 1
load 0
```

This would be a huge limitation when we trying to use this multiple times. Can I get some advice / work-around on this?
Thanks in advance.

Reply:
Using scratch slots seems very reasonable to me. I'm not trying to remove it, just to improve the OpCost. I hope ideas 
 to be helpful:

1. Add a method like `pyteal.TealBlock.NormalizeBlocks` to minimize the generated teal code like `store X; store Y; load Y; load X;`
2. Add a functionality like [yesql](https://www.npmjs.com/package/yesql) allowing users to write subroutines with teal code (with allowed op codes from pyteal.Op) directly.

Reply:
@PabloLION Let me know if the note helps to address your question + resolves the thread?

* Double checking:  Have you seen https://pyteal.readthedocs.io/en/stable/compiler_optimization.html?  We've considered extending the capability though it's not at the top of our priority list.  
* I _think_ https://github.com/algorand/pyteal-utils#inline-assembly fulfills the desire for access to write TEAL source.

Reply:
I didn't know about the "Compiler Optimization", and the InlineAssembly is amazing! Thanks for the clarification🥰